dfu: img_util: Upload to slot0 if not running from slot0

nordicjm opened the pull request:
Fixes an issue for MCUboot DFU whereby images would always be uploaded to slot1 instead of slot0 meaning that updates would never take effect, but retains uploading to slot1 if the application is running from slot0

Fixes #91042

nordicjm committed:
dfu: img_util: Upload to slot0 if not running from slot0

Fixes an issue for MCUboot DFU whereby images would always be
uploaded to slot1 instead of slot0 meaning that updates would
never take effect, but retains uploading to slot1 if the
application is running from slot0

Signed-off-by: Jamie McCrae <jamie.mccrae@nordicsemi.no>

nordicjm commented:
@ofirshe try this fix

ofirshe commented:
Hi @nordicjm,

Thanks for looking into the ticket. 

However, I initially thought the more robust approach would be to allow downloading images to both slot-0 and slot-1 in [usb_dfu.c](https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/usb/device/class/dfu/usb_dfu.c), rather than restricting downloads to slot-1 (if it exists) and only falling back to slot-0 otherwise.

To clarify, here's what I mentioned on Discord:
> I'm using USB DFU with MCUboot and observed that dfu-util allows uploading from both alt=0 (slot-0) and alt=1 (slot-1), but downloading is only permitted to alt=1. Any attempt to download to alt=0 is rejected by the DFU handler. Is this restriction intentional? The reason I ask is that when the bootloader fails to find a valid image in slot-0 (e.g., E: Unable to find bootable image), it triggers FIH_PANIC, leaving the system in a non-recoverable state unless I manually flash slot-0 using west flash. Even if I download a valid image to slot-1 via DFU, the bootloader still panics and doesn't attempt to copy it to slot-0.

That said, as you suggested, enabling CONFIG_BOOT_BOOTSTRAP could also address this issue by allowing the bootloader to recover from a broken slot-0 by copying from slot-1 automatically.

nordicjm committed:
tests: dfu: img_util: Update for new placement and new test

Updates to use slot1 as the default upload location, also adds a
new sub-test which checks uploading goes to slot0 when running
from slot1

Signed-off-by: Jamie McCrae <jamie.mccrae@nordicsemi.no>

nordicjm commented:
> Hi @nordicjm,
> 
> Thanks for looking into the ticket.
> 
> However, I initially thought the more robust approach would be to allow downloading images to both slot-0 and slot-1 in [usb_dfu.c](https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/usb/device/class/dfu/usb_dfu.c), rather than restricting downloads to slot-1 (if it exists) and only falling back to slot-0 otherwise.
> 
> To clarify, here's what I mentioned on Discord:
> 
> > I'm using USB DFU with MCUboot and observed that dfu-util allows uploading from both alt=0 (slot-0) and alt=1 (slot-1), but downloading is only permitted to alt=1. Any attempt to download to alt=0 is rejected by the DFU handler. Is this restriction intentional? The reason I ask is that when the bootloader fails to find a valid image in slot-0 (e.g., E: Unable to find bootable image), it triggers FIH_PANIC, leaving the system in a non-recoverable state unless I manually flash slot-0 using west flash. Even if I download a valid image to slot-1 via DFU, the bootloader still panics and doesn't attempt to copy it to slot-0.
> 
> That said, as you suggested, enabling `CONFIG_BOOT_BOOTSTRAP` could also address this issue by allowing the bootloader to recover from a broken slot-0 by copying from slot-1 automatically.

You can only upload to 1 slot as that's how the API is designed (the function returns the flash area ID of where to upload to), so it would either be slot 0 or slot 1
